File: src/service-setting-service/src/repositories/cosmosClient.ts
import { CosmosClient, Container, Database } from '@azure/cosmos';
import { config } from '../config';
import logger from '../utils/logger';

export class CosmosDBClient {
  private client: CosmosClient;
  private database: Database | null = null;
  private container: Container | null = null;

  constructor() {
    this.client = new CosmosClient({
      endpoint: config.cosmos.endpoint,
      key: config.cosmos.key,
    });
  }

  async initialize(): Promise<void> {
    try {
      const { database } = await this.client.databases.createIfNotExists({
        id: config.cosmos.database,
      });
      this.database = database;

      const { container } = await database.containers.createIfNotExists({
        id: config.cosmos.container,
        partitionKey: { paths: ['/tenantId'] },
      });
      this.container = container;

      logger.info('CosmosDB initialized successfully');
    } catch (error) {
      logger.error('Failed to initialize CosmosDB:', error);
      throw error;
    }
  }

  getContainer(): Container {
    if (!this.container) {
      throw new Error('CosmosDB container not initialized');
    }
    return this.container;
  }

  getDatabase(): Database {
    if (!this.database) {
      throw new Error('CosmosDB database not initialized');
    }
    return this.database;
  }
}

export const cosmosDBClient = new CosmosDBClient();

File: src/service-setting-service/app/repositories/cosmos_client.py
from azure.cosmos import CosmosClient, ContainerProxy, DatabaseProxy
from app.core.config import settings
from app.core.logger import logger


class CosmosDBClient:
    def __init__(self) -> None:
        self.client = CosmosClient(settings.COSMOS_ENDPOINT, settings.COSMOS_KEY)
        self.database: DatabaseProxy | None = None
        self.container: ContainerProxy | None = None

    async def initialize(self) -> None:
        """Initialize CosmosDB database and container"""
        try:
            self.database = self.client.create_database_if_not_exists(id=settings.COSMOS_DATABASE)
            self.container = self.database.create_container_if_not_exists(
                id=settings.COSMOS_CONTAINER, partition_key={"paths": ["/tenant_id"]}
            )
            logger.info("CosmosDB initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize CosmosDB: {e}")
            raise

    def get_container(self) -> ContainerProxy:
        """Get CosmosDB container"""
        if not self.container:
            raise RuntimeError("CosmosDB container not initialized")
        return self.container


cosmos_client = CosmosDBClient()
